src/api/OnmsHTTPOptions.ts

```diff
@@ -1,118 +1,338 @@
+   */
+  public data(data?: any) {
+    this._data = data;
+    return this;
+  }
+}
+// tslint:enable:completed-docs variable-name whitespace
+
 /**
  * Options to be used when making HTTP ReST calls.
  * @module OnmsHTTPOptions
  */
 export class OnmsHTTPOptions {
-  /** How long to wait for ReST calls to time out. */
-  public get timeout(): number {
-    if (this[TIMEOUT_PROP]) {
-      return this[TIMEOUT_PROP];
-    }
-    return DEFAULT_TIMEOUT;
+  /**
+   * Create a new builder for an [[OnmsHTTPOptions]] object.
+   * @param options if an existing options object is passed, the builder will be pre-populated
+   */
+  public static newBuilder(options?: OnmsHTTPOptions) {
+    return new OnmsHTTPOptionsBuilder(options);
   }
 
-  public set timeout(t: number) {
-    this[TIMEOUT_PROP] = t;
+  /** How long to wait for ReST calls to time out. */
+  public get timeout(): number | undefined {
+    return this[TIMEOUT_PROP] || undefined;
   }
 
   /** The authentication config that should be used. */
-  public get auth(): OnmsAuthConfig {
-    if (this[AUTH_PROP]) {
-      return this[AUTH_PROP];
+  public get auth(): OnmsAuthConfig | undefined {
+    const auth = this[AUTH_PROP];
+    if (auth !== null && auth !== undefined) {
+      return auth;
     }
     if (this.server && this.server.auth) {
       return this.server.auth;
     }
-    return {} as OnmsAuthConfig;
-  }
-
-  public set auth(a: OnmsAuthConfig) {
-    this[AUTH_PROP] = a;
+    return undefined;
   }
 
-  /** The server to use if no server is set on the HTTP implementation. */
-  public server?: OnmsServer | null;
+  /** The server to use instead of that provided by the HTTP implementation. */
+  public readonly server: OnmsServer | null;
 
   /** HTTP headers to be passed to the request. */
-  public headers = {} as IHash<string>;
+  public readonly headers = {} as IHash<string>;
 
   /** HTTP parameters to be passed on the URL. */
-  public parameters = {} as IHash<string | string[]>;
+  public readonly parameters = {} as IHash<string | string[]>;
 
   /** HTTP data to be passed when POSTing */
-  public data: any;
+  public readonly data: any;
 
   /**
    * The default timeout associated with these options.
    *
    * This is a trick for making sure serialization to JSON happens properly
    * without exposing internals.
    */
-  private [TIMEOUT_PROP]: number;
+  private readonly [TIMEOUT_PROP]: number | null;
 
   /**
    * The default authentication credentials associated with these options.
    *
    * This is a trick for making sure serialization to JSON happens properly
    * without exposing internals.
    */
-  private [AUTH_PROP]: OnmsAuthConfig;
+  private [AUTH_PROP]: OnmsAuthConfig | null;
 
   /**
    * Construct a new OnmsHTTPOptions object.
    * @constructor
    */
-  constructor(timeout?: number, auth?: OnmsAuthConfig, server?: OnmsServer) {
-    this.timeout = timeout || DEFAULT_TIMEOUT;
-    this.auth = auth || new OnmsAuthConfig();
+  constructor(
```

Comment:
The constructor should now take a single argument, the builder, and the member properties should then be derived from the builder.

Reply:
...really?  I thought the whole point of the builder pattern is the *builder* makes the object. Feels like data envy to me.

Reply:
The builder does build the object. I probably wasn't clear enough in my comment.

The builder's `build()` will look like this:
```
return new OnmsFoo(this);
```

In java we actually make the constructor of the immutable object so you can only build it through the builder (not sure if you can do that in TS). Either way it cleans up the signature of the constructor and ensures the object is being created by using the builder.

Here is an example in TS:
https://medium.com/@itayelgazar/the-builder-pattern-in-node-js-typescript-4b81a70b2ea5

And for reference here is how we are doing it in the integration API:
https://github.com/OpenNMS/opennms-integration-api/blob/c855ddfa6d9292211b4b859bb6d797167ed2c51f/common/src/main/java/org/opennms/integration/api/v1/model/immutables/ImmutableAlarm.java#L65

Reply:
Huh, ok. I guess I hadn't seen how you'd been doing it, traditionally I'd only ever seen it like they show [here](https://en.wikipedia.org/wiki/Builder_pattern#Java).

Reply:
In the case of our Java impls the builder is a nested class of the immutable object itself so I  in that case it makes more sense to take the builder as the only arg (and we are also specifically attempting to enforce construction via builder only).

In the example you linked the entity class and builder are not tightly coupled explicitly via inner class so it makes sense there to abide by the entity classes existing constructor.

Now how that translates to Typescript I'm not sure since the builder is not an inner class and is not owned by the other class. Also we can't enforce construction through the builder since I don't think we can make the constructor private.

That being said, I think this boils down to just being a nitpick and essentially personal preference. So I'll leave it up to you how you want to implement it. I'm fine with the existing multi-arg constructor.

Reply:
Yeah, I can see the other way working, just hadn't really seen it done before.  I went ahead and switched things around (tho' there's no real way to do the "private builder" thing that isn't hacky).